File: apps/manifold_api/routers/relations.py
# apps/manifold_api/routers/relations.py
"""Thought-to-thought relations endpoints."""
from fastapi import APIRouter, Depends, HTTPException
from libs.manifold_core.storage.qdrant_store import QdrantStore
from apps.manifold_api.dependencies import get_qdrant_store
from typing import Literal, Dict, Any, List, Set
from datetime import datetime
from pydantic import BaseModel

router = APIRouter(prefix="/v1/memory/thought", tags=["relations"])


class BatchLinkRelation(BaseModel):
    """Single relation in batch request."""
    related_id: str
    relation_type: Literal["supports", "contradicts", "followup", "duplicate", "related"] = "related"
    weight: float = 1.0
    description: str | None = None  # Optional explanation of why this relation exists


class BatchLinkRequest(BaseModel):
    """Batch link multiple relations."""
    relations: List[BatchLinkRelation]


@router.post("/{thought_id}/related")
def link_related(
    thought_id: str,
    body: dict,
    store: QdrantStore = Depends(get_qdrant_store),
):
    """Link thought to related thought."""
    related_id = body.get("related_id")
    if not related_id:
        raise HTTPException(400, "related_id required")
    relation_type: Literal["supports","contradicts","followup","duplicate","related"] = body.get("relation_type", "related")
    weight = body.get("weight", 1.0)
    description = body.get("description")  # Optional explanation of the relation
    
    # Fetch source
    source = store.get_by_id(thought_id)
    if not source:
        raise HTTPException(404, f"Thought {thought_id} not found")
    
    # Add to links.related_thoughts
    links = source.get("links", {})
    related = links.get("related_thoughts", [])
    if related_id not in related:
        related.append(related_id)
    links["related_thoughts"] = related

    # Maintain typed relations with timestamp
    now = datetime.utcnow().isoformat() + "Z"
    typed: List[Dict[str, Any]] = links.get("relations", [])
    if not any(r.get("related_id") == related_id for r in typed):
        relation_dict = {
            "related_id": related_id,
            "type": relation_type,
            "weight": float(weight),
            "created_at": now,
        }
        if description:
            relation_dict["description"] = description
        typed.append(relation_dict)
    links["relations"] = typed
    source["links"] = links
    
    # Update source's updated_at timestamp
    source["updated_at"] = now
    
    # Update
    store.upsert_point(thought_id, source, vectors={})
    
    return {"status": "linked", "created_at": now}


@router.post("/{thought_id}/related/batch")
def batch_link_related(
    thought_id: str,
    body: BatchLinkRequest,
    store: QdrantStore = Depends(get_qdrant_store),
):
    """Link multiple relations at once."""
    # Fetch source
    source = store.get_by_id(thought_id)
    if not source:
        raise HTTPException(404, f"Thought {thought_id} not found")
    
    # Initialize links if needed
    links = source.get("links", {})
    related = links.get("related_thoughts", [])
    typed: List[Dict[str, Any]] = links.get("relations", [])
    
    now = datetime.utcnow().isoformat() + "Z"
    results = []
    linked_count = 0
    skipped_count = 0
    
    # Process each relation
    for relation in body.relations:
        related_id = relation.related_id
        relation_type = relation.relation_type
        weight = relation.weight
        description = relation.description  # Optional explanation
        
        # Check if already linked
        already_in_related = related_id in related
        already_typed = any(r.get("related_id") == related_id for r in typed)
        
        if already_in_related and already_typed:
            skipped_count += 1
            results.append({
                "related_id": related_id,
                "status": "skipped",
                "reason": "already_linked"
            })
            continue
        
        # Add to related_thoughts if not present
        if not already_in_related:
            related.append(related_id)
        
        # Add to typed relations if not present
        if not already_typed:
            relation_dict = {
                "related_id": related_id,
                "type": relation_type,
                "weight": float(weight),
                "created_at": now,
            }
            if description:
                relation_dict["description"] = description
            typed.append(relation_dict)
            linked_count += 1
            result = {
                "related_id": related_id,
                "status": "linked",
                "relation_type": relation_type,
                "weight": weight,
                "created_at": now
            }
            if description:
                result["description"] = description
            results.append(result)
        else:
            # Update existing typed relation
            for r in typed:
                if r.get("related_id") == related_id:
                    r["type"] = relation_type
                    r["weight"] = float(weight)
                    if description:
                        r["description"] = description
                    r["updated_at"] = now
                    linked_count += 1
                    result = {
                        "related_id": related_id,
                        "status": "updated",
                        "relation_type": relation_type,
                        "weight": weight,
                        "updated_at": now
                    }
                    if description:
                        result["description"] = description
                    results.append(result)
                    break
    
    # Update source
    links["related_thoughts"] = related
    links["relations"] = typed
    source["links"] = links
    source["updated_at"] = now
    
    # Update in store
    store.upsert_point(thought_id, source, vectors={})
    
    return {
        "status": "ok",
        "thought_id": thought_id,
        "linked_count": linked_count,
        "skipped_count": skipped_count,
        "total_relations": len(body.relations),
        "results": results,
        "created_at": now
    }


@router.delete("/{thought_id}/related/{related_id}")
def unlink_related(
    thought_id: str,
    related_id: str,
    store: QdrantStore = Depends(get_qdrant_store),
):
    """Unlink related thought."""
    source = store.get_by_id(thought_id)
    if not source:
        raise HTTPException(404, f"Thought {thought_id} not found")
    
    links = source.get("links", {})
    related = links.get("related_thoughts", [])
    if related_id in related:
        related.remove(related_id)
    links["related_thoughts"] = related
    # Remove from typed list as well
    typed: List[Dict[str, Any]] = links.get("relations", [])
    typed = [r for r in typed if r.get("related_id") != related_id]
    links["relations"] = typed
    source["links"] = links
    
    store.upsert_point(thought_id, source, vectors={})
    
    return {"status": "unlinked"}


@router.get("/{thought_id}/related")
def get_related(
    thought_id: str,
    depth: int = 1,
    store: QdrantStore = Depends(get_qdrant_store),
):
    """Get related thoughts up to depth."""
    source = store.get_by_id(thought_id)
    if not source:
        raise HTTPException(404, f"Thought {thought_id} not found")
    
    # Direct outgoing links from this thought
    outgoing_ids = source.get("links", {}).get("related_thoughts", [])
    outgoing = [{"from_id": thought_id, "to_id": rid, "relation_type": "related"} for rid in outgoing_ids]
    # Typed outgoing
    typed_outgoing = []
    for r in source.get("links", {}).get("relations", []) or []:
        rel_dict = {
            "from_id": thought_id,
            "to_id": r.get("related_id"),
            "relation_type": r.get("type", "related"),
            "weight": r.get("weight", 1.0),
        }
        if r.get("description"):
            rel_dict["description"] = r.get("description")
        typed_outgoing.append(rel_dict)
    
    # Find incoming links (scan all thoughts that link to this one)
    # CRITICAL: Only scan active thoughts (exclude deleted)
    incoming = []
    typed_incoming = []
    all_thoughts = store.scroll(
        payload_filter={"must": [{"key": "status", "match": {"value": "active"}}]},
        limit=10000
    )
    for point in all_thoughts:
        # Double-check: skip deleted (shouldn't happen with filter, but be safe)
        if point.payload.get("status") == "deleted":
            continue
        related_ids = point.payload.get("links", {}).get("related_thoughts", [])
        if thought_id in related_ids:
            incoming.append({"from_id": point.id, "to_id": thought_id, "relation_type": "related"})
        # typed incoming
        for r in point.payload.get("links", {}).get("relations", []) or []:
            if r.get("related_id") == thought_id:
                rel_dict = {
                    "from_id": point.id,
                    "to_id": thought_id,
                    "relation_type": r.get("type", "related"),
                    "weight": r.get("weight", 1.0),
                }
                if r.get("description"):
                    rel_dict["description"] = r.get("description")
                typed_incoming.append(rel_dict)
    
    # Get full thought objects
    related_ids: Set[str] = set(outgoing_ids)
    related_ids.update([inc["from_id"] for inc in incoming])
    related_ids.update([e["to_id"] for e in typed_outgoing if e.get("to_id")])
    related_ids.update([e["from_id"] for e in typed_incoming if e.get("from_id")])
    
    thoughts = []
    for rid in related_ids:
        thought = store.get_by_id(rid)
        if thought:
            # CRITICAL: Skip deleted thoughts (shouldn't be in workflows)
            if thought.get("status") == "deleted":
                continue
            # include id in payload for frontend convenience
            if "id" not in thought:
                try:
                    thought["id"] = rid
                except Exception:
                    pass
            thoughts.append(thought)
    
    # include id on source as well
    if "id" not in source:
        try:
            source["id"] = thought_id
        except Exception:
            pass

    return {
        "status": "ok",
        "thought": source,
        "thoughts": thoughts,
        "incoming": incoming,
        "outgoing": outgoing,
        "typed_incoming": typed_incoming,
        "typed_outgoing": typed_outgoing,
    }


@router.get("/{thought_id}/related/facets")
def get_related_facets(
    thought_id: str,
    store: QdrantStore = Depends(get_qdrant_store),
):
    """Facet counts among neighbors (type, status, tickers, sectors)."""
    rel = get_related(thought_id, store=store)  # type: ignore
    neighbor_ids = set()
    for e in rel["incoming"]:
        neighbor_ids.add(e["from_id"])
    for e in rel["outgoing"]:
        neighbor_ids.add(e["to_id"])
    for e in rel.get("typed_incoming", []):
        neighbor_ids.add(e["from_id"])
    for e in rel.get("typed_outgoing", []):
        neighbor_ids.add(e["to_id"])
    neighbors = []
    for nid in neighbor_ids:
        t = store.get_by_id(nid)
        if t:
            neighbors.append(t)
    def add_count(d: Dict[str,int], key: str):
        d[key] = d.get(key, 0) + 1
    facets = {"type": {}, "status": {}, "tickers": {}, "sectors": {}}
    for p in neighbors:
        add_count(facets["type"], p.get("type","unknown"))
        add_count(facets["status"], p.get("status","unknown"))
        for x in p.get("tickers", []) or []:
            add_count(facets["tickers"], x)
        for x in p.get("sectors", []) or []:
            add_count(facets["sectors"], x)
    # format
    formatted = {k: [{"value": kk, "count": vv} for kk, vv in v.items()] for k, v in facets.items()}
    return {"status": "ok", "facets": formatted}


@router.get("/{thought_id}/related/graph")
def get_related_graph(
    thought_id: str,
    depth: int = 1,
    store: QdrantStore = Depends(get_qdrant_store),
):
    """Return simple nodes/edges graph up to depth."""
    if depth < 1:
        depth = 1
    visited: Set[str] = set()
    frontier: Set[str] = {thought_id}
    nodes: Dict[str, Any] = {}
    edges: List[Dict[str, Any]] = []
    for _ in range(min(depth, 3)):
        next_frontier: Set[str] = set()
        for nid in list(frontier):
            if nid in visited:
                continue
            visited.add(nid)
            t = store.get_by_id(nid)
            if not t:
                continue
            nodes[nid] = t
            links = t.get("links", {})
            for rid in links.get("related_thoughts", []) or []:
                edges.append({"from": nid, "to": rid, "type": "related", "weight": 1.0})
                next_frontier.add(rid)
            for r in links.get("relations", []) or []:
                rid = r.get("related_id")
                if rid:
                    edge_dict = {"from": nid, "to": rid, "type": r.get("type","related"), "weight": r.get("weight",1.0)}
                    if r.get("description"):
                        edge_dict["description"] = r.get("description")
                    edges.append(edge_dict)
                    next_frontier.add(rid)
        frontier = next_frontier
    return {
        "status": "ok",
        "nodes": [{"id": k, "payload": v} for k, v in nodes.items()],
        "edges": edges,
    }


@router.get("/{thought_id}/tree")
def get_thought_tree(
    thought_id: str,
    depth: int = 2,
    store: QdrantStore = Depends(get_qdrant_store),
):
    """Get hierarchical tree: parent → thought → children, plus related thoughts."""
    source = store.get_by_id(thought_id)
    if not source:
        raise HTTPException(404, f"Thought {thought_id} not found")
    
    result = {
        "status": "ok",
        "root": {"id": thought_id, "payload": source},
        "parent": None,
        "children": [],
        "related": [],
    }
    
    # Get parent (if exists)
    parent_id = source.get("parent_id")
    if parent_id:
        parent = store.get_by_id(parent_id)
        if parent:
            result["parent"] = {"id": parent_id, "payload": parent}
    
    # Get children
    # CRITICAL: Only get active children (exclude deleted)
    children_filter = {
        "must": [
            {"key": "parent_id", "match": {"value": thought_id}},
            {"key": "status", "match": {"value": "active"}}
        ]
    }
    children_points = store.scroll(payload_filter=children_filter, limit=100)
    
    children = []
    for p in children_points:
        # Double-check: skip deleted (shouldn't happen with filter, but be safe)
        if p.payload.get("status") == "deleted":
            continue
        children.append({
            "id": str(p.id),
            "payload": p.payload,
            "ordinal": p.payload.get("ordinal", 0)
        })
    
    children.sort(key=lambda x: x["ordinal"])
    result["children"] = children
    
    # Get related thoughts
    links = source.get("links", {})
    related_ids = links.get("related_thoughts", [])
    related = []
    for rid in related_ids:
        thought = store.get_by_id(rid)
        if thought:
            # CRITICAL: Skip deleted thoughts (shouldn't be in workflows)
            if thought.get("status") == "deleted":
                continue
            related.append({"id": rid, "payload": thought})
    
    result["related"] = related
    
    return result


@router.get("/with-relation-type/{relation_type}")
def get_thoughts_with_relation_type(
    relation_type: Literal["supports", "contradicts", "followup", "duplicate", "related"],
    limit: int = 20,  # Reduced from 100 for token efficiency
    include_thoughts: bool = True,
    include_content: bool = False,  # Default False to save tokens (only applies if include_thoughts=True)
    from_dt: str | None = None,
    to_dt: str | None = None,
    days: int | None = None,
    session_id: str | None = None,
    workspace_id: str | None = None,
    mcp: bool = False,  # If True, apply token safety limits (MCP/Agent calls)
    store: QdrantStore = Depends(get_qdrant_store),
):
    """Get all thoughts that have at least one relation of the specified type. Hard max limit of 100 for token safety (only if mcp=true).
    
    Returns pairs of thoughts connected by the specified relation type.
    If include_thoughts=True (default), returns full thought objects instead of just IDs.
    
    Supports temporal filtering via from_dt/to_dt (ISO format) or days (relative from now).
    Filters apply to the source thought's created_at timestamp.
    """
    # Cap limit for MCP calls (token safety), frontend can use higher limits
    if mcp and limit > 100:
        limit = 100
    from datetime import datetime, timedelta
    
    # Determine effective date range
    eff_from = from_dt
    eff_to = to_dt
    if not eff_from and not eff_to and days and days > 0:
        now = datetime.utcnow()
        eff_from = (now - timedelta(days=days)).isoformat() + "Z"
        eff_to = now.isoformat() + "Z"
    
    # Build filter
    # CRITICAL: Exclude deleted thoughts from discovery workflows
    must = []
    must.append({"key": "status", "match": {"value": "active"}})  # Only active thoughts
    if session_id:
        must.append({"key": "session_id", "match": {"value": session_id}})
    if workspace_id:
        must.append({"key": "workspace_id", "match": {"value": workspace_id}})
    if eff_from:
        must.append({"key": "created_at", "range": {"gte": eff_from}})
    if eff_to:
        must.append({"key": "created_at", "range": {"lte": eff_to}})
    
    filters = {"must": must} if must else None
    all_thoughts = store.scroll(payload_filter=filters, limit=10000)
    pairs = []
    thought_cache = {}  # Cache thoughts to avoid multiple lookups
    
    for point in all_thoughts:
        # Double-check: skip deleted (shouldn't happen with filter, but be safe)
        if point.payload.get("status") == "deleted":
            continue
        thought_id = str(point.id)
        if thought_id not in thought_cache:
            thought_cache[thought_id] = point.payload
        
        links = point.payload.get("links", {})
        relations = links.get("relations", []) or []
        
        for r in relations:
            if r.get("type") == relation_type:
                related_id = r.get("related_id")
                if related_id:
                    # Cache related thought if not already cached
                    if related_id not in thought_cache:
                        related_thought = store.get_by_id(related_id)
                        if related_thought:
                            # CRITICAL: Skip deleted thoughts (shouldn't be in workflows)
                            if related_thought.get("status") == "deleted":
                                continue
                            thought_cache[related_id] = related_thought
                        else:
                            continue
                    
                    # Double-check related thought is not deleted
                    if related_id in thought_cache and thought_cache[related_id].get("status") == "deleted":
                        continue
                    
                    pair = {
                        "from_id": thought_id,
                        "to_id": related_id,
                        "relation_type": relation_type,
                        "weight": r.get("weight", 1.0),
                        "created_at": r.get("created_at"),
                    }
                    if r.get("description"):
                        pair["description"] = r.get("description")
                    
                    # Include full thoughts if requested
                    if include_thoughts:
                        from_thought = thought_cache.get(thought_id, {})
                        to_thought = thought_cache.get(related_id, {})
                        
                        # Strip content if include_content=False
                        if not include_content:
                            def _prune_thought(thought: dict) -> dict:
                                """Keep only essential fields for relation-type queries."""
                                return {
                                    "id": thought.get("id"),
                                    "title": thought.get("title"),
                                    "summary": thought.get("summary"),
                                    "type": thought.get("type"),
                                    "tags": thought.get("tags", []),
                                    "tickers": thought.get("tickers", []),
                                    "confidence_score": thought.get("confidence_score"),
                                    "created_at": thought.get("created_at"),
                                    "status": thought.get("status"),
                                }
                            from_thought = _prune_thought(from_thought) if from_thought else {}
                            to_thought = _prune_thought(to_thought) if to_thought else {}
                        
                        pair["from_thought"] = from_thought
                        pair["to_thought"] = to_thought
                    
                    pairs.append(pair)
    
    # Deduplicate (bidirectional relations might appear twice)
    seen = set()
    unique_pairs = []
    for pair in pairs:
        key = tuple(sorted([pair["from_id"], pair["to_id"]]))
        if key not in seen:
            seen.add(key)
            unique_pairs.append(pair)
    
    return {
        "status": "ok",
        "relation_type": relation_type,
        "count": len(unique_pairs),
        "pairs": unique_pairs
    }

File: mcps/manifold/src/lib/tools/relations.ts
import { z } from 'zod';
import { callManifold } from '../api-client.js';
import { RelationPayloadSchema, BatchLinkRequestSchema } from '../schemas.js';

export const linkRelatedTool = {
  name: 'mf-link-related',
  config: {
    title: 'Link Related Thought',
    description: 'Create a single typed relation from one thought to another. Relation types: supports (thought B supports claim A), contradicts (thought B contradicts A), followup (B is a follow-up to A), duplicate (B is duplicate of A), related (general connection). Weight (0-1) indicates relation strength. Optional description field explains WHY the relation exists (e.g., "Both discuss same regulatory change", "Thought B provides evidence for A\'s hypothesis"). Always provide description when creating relations to make connections transparent. For creating multiple relations efficiently, use mf-batch-link-related instead. Relations are bidirectional for querying but stored directionally. Returns creation timestamp.',
    inputSchema: z.object({ 
      thought_id: z.string().describe('Source thought ID (the thought creating the relation).'),
      payload: RelationPayloadSchema.describe('Relation payload containing related_id, relation_type, optional weight (0-1, default 1.0), and optional description explaining why this relation exists.')
    }).shape
  },
  handler: async (input: { thought_id: string; payload: z.infer<typeof RelationPayloadSchema> }) => {
    const res = await callManifold(`/v1/memory/thought/${input.thought_id}/related`, { method: 'POST', body: JSON.stringify(input.payload) }, 10000);
    return { content: [{ type: 'text', text: JSON.stringify(res, null, 2) }] };
  }
};

export const batchLinkRelatedTool = {
  name: 'mf-batch-link-related',
  config: {
    title: 'Batch Link Related Thoughts',
    description: 'Create multiple relations from one thought to multiple other thoughts in a single API call. Much more efficient than multiple mf-link-related calls. Supports up to 100 relations per call. Each relation can have its own type (supports, contradicts, followup, duplicate, related), weight (0-1), and optional description explaining WHY the relation exists. Always provide descriptions to make connections transparent (e.g., "Both discuss same regulatory change", "Thought B provides evidence for A\'s hypothesis"). Use this when linking a thought to multiple similar thoughts, creating a network of connections, or establishing complex relationship structures. Returns detailed results for each relation (linked, skipped, updated).',
    inputSchema: z.object({ 
      thought_id: z.string().describe('Source thought ID (the thought creating all relations).'),
      relations: BatchLinkRequestSchema.shape.relations.describe('Array of 1-100 relation objects. Each must have related_id, optional relation_type (default: "related"), optional weight (default: 1.0), and optional description explaining why the relation exists.')
    }).shape
  },
  handler: async (input: { thought_id: string; relations: z.infer<typeof BatchLinkRequestSchema>['relations'] }) => {
    const res = await callManifold(
      `/v1/memory/thought/${input.thought_id}/related/batch`, 
      { 
        method: 'POST', 
        body: JSON.stringify({ relations: input.relations }) 
      }, 
      15000
    );
    return { content: [{ type: 'text', text: JSON.stringify(res, null, 2) }] };
  }
};

export const unlinkRelatedTool = {
  name: 'mf-unlink-related',
  config: {
    title: 'Unlink Related Thought',
    description: 'Remove a relation between two thoughts. Deletes the relation from both the related_thoughts list and the typed relations array. Use this to correct mistaken relations, remove false "duplicate" links after verification, or clean up relations before deleting a thought. Returns status confirming unlinking. Note: This removes the relation entirely - if you want to change the relation type or weight, use mf-batch-link-related with update.',
    inputSchema: z.object({ 
      thought_id: z.string().describe('Source thought ID (the thought removing the relation).'),
      related_id: z.string().describe('Target thought ID (the thought to unlink from source).')
    }).shape
  },
  handler: async (input: { thought_id: string; related_id: string }) => {
    const res = await callManifold(`/v1/memory/thought/${input.thought_id}/related/${input.related_id}`, { method: 'DELETE' }, 10000);
    return { content: [{ type: 'text', text: JSON.stringify(res, null, 2) }] };
  }
};

export const getRelatedTool = {
  name: 'mf-get-related',
  config: {
    title: 'Get Related Thoughts',
    description: 'Get all thoughts related to a given thought, including both outgoing (thought → others) and incoming (others → thought) relations. Returns full thought objects, typed relations (supports, contradicts, followup, duplicate, related), and relation metadata. Depth parameter controls how many hops to traverse (1 = direct neighbors only, 2 = neighbors of neighbors, max 3). Use this to understand the thought\'s position in the knowledge graph, explore connections, or analyze relationship patterns. Returns structured data with incoming/outgoing relations separately identified.',
    inputSchema: z.object({ 
      thought_id: z.string().describe('The thought ID to get relations for.'),
      depth: z.number().int().min(1).max(3).default(1).describe('Traversal depth: 1 = direct neighbors only, 2 = 2 hops, 3 = 3 hops (max). Higher depth can be slow for highly connected thoughts.')
    }).shape
  },
  handler: async (input: { thought_id: string; depth?: number }) => {
    const params = new URLSearchParams();
    if (input.depth) params.append('depth', String(input.depth));
    const res = await callManifold(`/v1/memory/thought/${input.thought_id}/related?${params.toString()}`, {}, 15000);
    return { content: [{ type: 'text', text: JSON.stringify(res, null, 2) }] };
  }
};

export const getRelatedFacetsTool = {
  name: 'mf-related-facets',
  config: {
    title: 'Related Facets',
    description: 'Get facet/aggregation counts for all thoughts related to a given thought. Returns counts by type (observation, hypothesis, analysis, etc.), status (active, deleted, quarantined), tickers mentioned, and sectors. Useful for understanding the context around a thought - what types of thoughts connect to it, which tickers are most prominent in its neighborhood, what sectors are represented. Helps with discovery, pattern analysis, and understanding thought clusters.',
    inputSchema: z.object({ 
      thought_id: z.string().describe('The thought ID to analyze neighbors of.')
    }).shape
  },
  handler: async (input: { thought_id: string }) => {
    const res = await callManifold(`/v1/memory/thought/${input.thought_id}/related/facets`, {}, 10000);
    return { content: [{ type: 'text', text: JSON.stringify(res, null, 2) }] };
  }
};

export const getRelatedGraphTool = {
  name: 'mf-related-graph',
  config: {
    title: 'Related Graph',
    description: 'Get graph structure (nodes and edges) for the subgraph around a thought. Returns arrays of nodes (thoughts) and edges (relations) with their types and weights. Useful for visualization, graph analysis, understanding network topology, or exporting subgraphs. Depth controls traversal (1-3 hops). Each edge includes relation type (supports, contradicts, followup, duplicate, related) and weight. More efficient than get-related when you only need structure, not full thought content.',
    inputSchema: z.object({ 
      thought_id: z.string().describe('Central thought ID for the subgraph.'),
      depth: z.number().int().min(1).max(3).default(1).describe('Graph depth: 1 = direct neighbors, 2-3 = extended neighborhood.')
    }).shape
  },
  handler: async (input: { thought_id: string; depth?: number }) => {
    const params = new URLSearchParams();
    if (input.depth) params.append('depth', String(input.depth));
    const res = await callManifold(`/v1/memory/thought/${input.thought_id}/related/graph?${params.toString()}`, {}, 15000);
    return { content: [{ type: 'text', text: JSON.stringify(res, null, 2) }] };
  }
};

export const getThoughtTreeTool = {
  name: 'mf-get-thought-tree',
  config: {
    title: 'Get Thought Tree',
    description: 'Get hierarchical tree structure for a thought: parent (if exists), the thought itself, all children (sorted by ordinal), and related thoughts. Useful for understanding hierarchical structures where thoughts have parent-child relationships (e.g., analyses with numbered sub-points, decision trees, structured breakdowns). Returns full thought objects for parent, children (ordered), and related thoughts separately. Different from get-related-graph as it focuses on hierarchy (parent/children) plus relations, not just the relation network.',
    inputSchema: z.object({ 
      thought_id: z.string().describe('Root thought ID for the tree.'),
      depth: z.number().int().min(1).max(3).default(2).optional().describe('Maximum depth for related thoughts exploration. Default 2.')
    }).shape
  },
  handler: async (input: { thought_id: string; depth?: number }) => {
    try {
      const params = new URLSearchParams();
      if (input.depth) params.append('depth', String(input.depth));
      const res = await callManifold(`/v1/memory/thought/${input.thought_id}/tree?${params.toString()}`, {}, 15000);
      return { content: [{ type: 'text', text: JSON.stringify(res, null, 2) }] };
    } catch (e: any) {
      return { content: [{ type: 'text', text: `Error: ${e.message}` }], isError: true };
    }
  }
};

export const getThoughtsWithRelationTypeTool = {
  name: 'mf-get-thoughts-with-relation-type',
  config: {
    title: 'Get Thoughts with Relation Type',
    description: 'Find all thought pairs connected by a specific relation type across the entire knowledge graph. Useful for finding all "supports", "contradicts", "followup", "related", or "duplicate" relations. **KRITISCH - TOKEN-KOSTEN**: Default include_content=false (saves massive tokens!). Returns pairs with thought objects (from_thought, to_thought) if include_thoughts=true. Each pair includes relation weight and creation timestamp. Supports temporal filtering (from_dt/to_dt/days) to focus on recent thoughts. Much more efficient than scanning all thoughts manually. Use include_thoughts=false if you only need IDs and relation metadata. NOTE: For duplicate cleanup workflows, use mf-get-all-duplicates instead - it combines marked duplicate relations AND similarity-based detection.',
    inputSchema: z.object({
      relation_type: z.enum(['supports', 'contradicts', 'followup', 'duplicate', 'related']).describe('The relation type to filter by. Most common use: "duplicate" to find all duplicate pairs for cleanup.'),
      limit: z.number().int().min(1).max(100).default(20).describe('KRITISCH: Maximum number of pairs to return. Default 20 (token-efficient), max 100. Each pair with thoughts costs ~200-400 tokens without content, ~1000-4000 with!'),
      include_thoughts: z.boolean().default(true).optional().describe('If true (default): returns thought objects for both from_thought and to_thought. If false: returns only IDs and relation metadata. Set false to save tokens when you only need to identify pairs.'),
      include_content: z.boolean().default(false).optional().describe('KRITISCH: If false (default) and include_thoughts=true: returns only metadata (id, title, summary, type, tags, tickers, confidence_score, created_at, status) without full content/links/epistemology - MUCH more token-efficient! Set to true ONLY when you need full content. Default false saves massive tokens (80-90% reduction).'),
      from_dt: z.string().optional().describe('Start date in ISO format (e.g., "2025-11-01T00:00:00Z"). Use with to_dt for absolute date range. Filters by source thought created_at.'),
      to_dt: z.string().optional().describe('End date in ISO format. Use with from_dt for absolute date range. Filters by source thought created_at.'),
      days: z.number().int().min(1).optional().describe('Relative days from now (e.g., 7 = last week). Alternative to from_dt/to_dt.'),
      session_id: z.string().optional().describe('Filter to specific session.'),
      workspace_id: z.string().optional().describe('Filter to specific workspace.')
    }).shape
  },
  handler: async (input: { relation_type: string; limit?: number; include_thoughts?: boolean; include_content?: boolean; from_dt?: string; to_dt?: string; days?: number; session_id?: string; workspace_id?: string }) => {
    const params = new URLSearchParams();
    if (input.limit) params.append('limit', String(input.limit));
    if (input.include_thoughts !== undefined) params.append('include_thoughts', String(input.include_thoughts));
    if (typeof input.include_content === 'boolean') params.append('include_content', String(input.include_content));
    if (input.from_dt) params.append('from_dt', input.from_dt);
    if (input.to_dt) params.append('to_dt', input.to_dt);
    if (input.days) params.append('days', String(input.days));
    if (input.session_id) params.append('session_id', input.session_id);
    if (input.workspace_id) params.append('workspace_id', input.workspace_id);
    params.append('mcp', 'true');  // Always set mcp=true for MCP calls (token safety limits)
    const res = await callManifold(`/v1/memory/thought/with-relation-type/${input.relation_type}?${params.toString()}`, {}, 15000);
    
    // Additional MCP-level pruning: Even if backend returns content, ensure we only return essential fields
    // This ensures we always return token-efficient responses for relation-type queries (often used for discovery)
    if (res.pairs && Array.isArray(res.pairs) && input.include_thoughts && !input.include_content) {
      res.pairs = res.pairs.map((pair: any) => {
        const pruneThought = (thought: any) => {
          if (!thought) return thought;
          return {
            id: thought.id,
            title: thought.title,
            summary: thought.summary,
            type: thought.type,
            tags: thought.tags,
            tickers: thought.tickers,
            confidence_score: thought.confidence_score,
            created_at: thought.created_at,
            status: thought.status,
          };
        };
        return {
          ...pair,
          from_thought: pruneThought(pair.from_thought),
          to_thought: pruneThought(pair.to_thought),
        };
      });
    }
    
    return { content: [{ type: 'text', text: JSON.stringify(res, null, 2) }] };
  }
};


